GEODE-7283: Store MethodAuthorizer in Context

jujoramos opened the pull request:
The MethodInvocationAuthorizer should be unique per query and should
not be changed for a particular query while it's being executed so,
starting with this commit, it is directly stored as part of the
ExecutionContext. This also prevents the creation of multiple instances
of the DefaultQueryService and the configured authorizer itself while
the intermediate result objects are being traversed during the query
evaluation.

- Added tests.
- Fixed minor warnings.
- Replaced usages of 'junit.Assert' by 'assertj'.
- The AttributeDescriptor and MethodDispatch classes do not require
  MethodInvocationAuthorizer as a constructor parameter anymore, the
  authorizer is obtained from the context, and **only** when needed.
- The MethodDispatch class does not internally store and cache the
  MethodInvocationAuthorizer instance anymore, thus allowing Geode to
  change the authorizer in runtime without needing to re-load every
  already known method (expensive due to reflection).
- The ExecutionContext uses a no-op authorizer when created from a
  client cache: authorization and authentication logic happens only on
  server side, and the authorizer will be obtained from the cluster
  configuration service directly instead of the query service when OQL
  security is fully implemented.

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [X] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [X] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [X] Is your initial contribution a single, squashed commit?

- [X] Does `gradlew build` run cleanly?

- [X] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

jujoramos commented:
@DonalEvans @BenjaminPerryRoss

jhuynh1 commented on geode-core/src/integrationTest/java/org/apache/geode/cache/query/internal/ExecutionContextIntegrationTest.java:
```diff
@@ -0,0 +1,340 @@
+  }
+
+  @Test
+  public void testScopeIndex1() throws Exception {
```
is there a more descriptive name for this test?

jhuynh1 commented on geode-core/src/integrationTest/java/org/apache/geode/cache/query/internal/ExecutionContextIntegrationTest.java:
```diff
@@ -0,0 +1,340 @@
+  }
+
+  @Test
+  public void testMultiThreadedScopeIndex() {
```
Is there a more descriptive name that describes what the desired behavior is?

jhuynh1 commented on geode-core/src/integrationTest/java/org/apache/geode/cache/query/internal/ExecutionContextIntegrationTest.java:
```diff
@@ -0,0 +1,340 @@
+  }
+
+  @Test
+  public void testFunctionalAddToIndependentRuntimeItrMapWithoutIndex() throws Exception {
```
what is this testing (more descriptive name?)  It looks like we are checking that the internal id is the same as the index internal id?

jhuynh1 commented on geode-core/src/integrationTest/java/org/apache/geode/cache/query/internal/ExecutionContextIntegrationTest.java:
```diff
@@ -0,0 +1,340 @@
+  }
+
+  @Test
+  public void testObtainingRegionPath() throws Exception {
```
Is this confirming that the region path in the context matches after computing dependencies or in this case computingUltimateDependencies?  (Not knowing what the difference is at this time...)

jujoramos commented:
@DonalEvans: thanks for reviewing.
@jhuynh1: I didn't change the existing tests, just refactored the class to get rid of deprecated code and warnings, and added some new tests with meaningful names to test my changes. I'll dive into these particular tests to figure out what they're doing and change the names as requested, though, thanks.
